### ISSUE TITLE
Rethink scroll representation in Problem/Solution Theory

### ISSUE THREAD

**Issue by richardmarcus** [UFrameIT/UFrameIT#10]
We think we should improve the meta scroll* part. 
The current situation, where the meta annotations are on the problem theory are not ideal; since they include information about the solution theory. As Solution includes Problem, the annotation would be better on that. In particular, then a problem theory can be used in multiple scrolls. 

BUT that is still not ideal, since we cannot have multiple scrolls per solution theory. Think different verbalizations for different audiences. 

Let's be radical: I think a scroll should be an informal statement (in what theory?) or even an informal theory or maybe even an informal view, once we have them. cc: @Jazzpirate

**Comment by ComFreek** [comment]
> BUT that is still not ideal, since we cannot have multiple scrolls per solution theory.

We can since we can have arbitrarily many meta statements with the same key per theory, e.g.

```
theory SolutionThy =
  meta ?scrollMeta http://ProbThy1 "description1" ||
  meta ?scrollMeta http://ProbThy2 "description2" ||
|||
```

> I think a scroll should be an informal statement (in what theory?)

I'd second this -- what about **considering scroll pair definitions as facts**?

```
theory Meta =
  scroll: type ||
  scroll_ctor: URI -> URI -> string -> scroll || // prob URI, sol URI, description
|||

theory SituationThy : ?Meta =
  include ?Geometry ||
  scroll_fact1 | = scroll_ctor ?ProbThy ?SolutionTheory "..." ||
|||
```

Then, we have an even more uniform treatment of facts and scrolls:

- talking to the NPC in our current game would give you a scroll fact
- scrolls could give you scrolls again

It could look like

**Comment by bBoesl** [comment]
The second one, seems to be exactly what i had in mind,  looks promising

**Comment by kohlhase** [comment]
Ha, that looks good/interesting it certainly pays off to complain. 
BUT, technically I am not clear what `scroll_ctor` might be. It is intended to take two theories and make a fact. I have never seen such a beast. And it has always been an invariant that modules can contain declarations, but not the other way around.

**Comment by ComFreek** [comment]
> And it has always been an invariant that modules can contain declarations, but not the other way around.

That's true, dependency handling would be totally unaware of `scroll_ctor` applications.

Now that I think of it, all three

- the suggestion above,
- keys in `meta` statements,
- and diagram operators

all mention URIs to external, possibly non-included symbols. Interesting!

**Comment by bBoesl** [comment]
> Ha, that looks good/interesting it certainly pays off to complain.
> BUT, technically I am not clear what `scroll_ctor` might be. It is intended to take two theories and make a fact. I have never seen such a beast. And it has always been an invariant that modules can contain declarations, but not the other way around.

But is this not a natural step? 
For example:
Formalizing categories, where each object is a algebraic structure. Then it would make sense to have each object represented as its own theory and in the formalization of the category treat each of these theories as an object.
Correct me, if wrong =)

**Comment by kohlhase** [comment]
Yes,  it is a natural thing, but problematic from a language design point of view. 

One thing that helps is the mod operator, which takes theories and makes record types from them. That might help here as well. 
I am ccing @Jazzpirate for him to chime into this discussion.

**Comment by ComFreek** [comment]
For now we have settled on providing the scroll meta information in the solution theory.

E.g. see https://gl.mathhub.info/FrameIT/frameworld/-/blob/d3f14c97dc070c90421b95828c8c991571fb36ea/source/Scrolls/TriangleScrolls.mmt#L32-61:

```
theory AngleSum =
    theory AngleSum_Problem =
        include ?TriangleScroll_GeneralProblem ❙

        angleBAC
            : Σ α: ℝ. ⊦ ( ∠ B,A,C ) ≐ α ❘
            meta ?MetaAnnotations?description "Angle α at A"
        ❙

        angleABC
            : Σ β: ℝ. ⊦ ( ∠ A,B,C ) ≐ β ❘
            meta ?MetaAnnotations?label "Angle β at B"
        ❙
    ❚

    theory AngleSum_Solution =
        meta ?MetaAnnotations?label "AngleSum" ❙
        meta ?MetaAnnotations?problemTheory ?AngleSum/AngleSum_Problem ❙
        meta ?MetaAnnotations?solutionTheory ?AngleSum/AngleSum_Solution ❙
        meta ?MetaAnnotations?description "Given a triangle ABC and two known angles, we can deduce the missing angle by the sum of interior angles in triangles always being 180°" ❙

       include ?AngleSum/AngleSum_Problem ❙

        angleBCA
            : Σ γ: ℝ. ⊦ ( ∠ B,C,A ) ≐ γ ❘
            = ⟨180.0 - (πl angleBAC) - (πl angleABC), sketch "By sum of interior angles = 180° in triangles"⟩ ❘
            meta ?MetaAnnotations?description "The deduced angle by calculating 180° - α - β"
        ❙
    ❚
❚
```

> BUT that is still not ideal, since we cannot have multiple scrolls per solution theory. Think different verbalizations for different audiences.

Indeed. Ultimately, we might want a *many-to-many* table of (problem, solution) theory pairs.
Until that becomes a necessity, I'd rather not complicate the current setup anymore. Hence closing. Feel free to reopen if the issue resurfaces.

**Comment by ComFreek** [comment]
@richardmarcus I've again revisited this problem and revised how scrolls are to be formalized. See https://gl.mathhub.info/FrameIT/frameworld/-/tree/devel/source/Scrolls for how it works as of today.

TL;DR: scrolls are nothing but theories declaring two metadatums referencing a problem and a solution theory. As such, they are really flexible.
